src/main/java/com/github/junit5docker/DockerExtension.java

```diff
@@ -109,14 +88,29 @@ private String findImageName(Docker dockerAnnotation) {
     }
 
     @Override
-    public void afterAll(ExtensionContext containerExtensionContext) {
-        Docker dockerAnnotation = findDockerAnnotation(containerExtensionContext);
-        if (!dockerAnnotation.newForEachCase()) dockerClient.stopAndRemoveContainer(containerId);
+    public void afterAll(ExtensionContext context) {
+        forEachDocker(context, docker -> !docker.newForEachCase(), this::stopAndRemove);
     }
 
     @Override
     public void afterEach(ExtensionContext context) {
-        Docker dockerAnnotation = findDockerAnnotation(context);
-        if (dockerAnnotation.newForEachCase()) dockerClient.stopAndRemoveContainer(containerId);
+        forEachDocker(context, Docker::newForEachCase, this::stopAndRemove);
     }
+
+    private void stopAndRemove(ExtensionContext context, Docker docker) {
+        ContainerInfo containerInfo = getStore(context).remove(docker, ContainerInfo.class);
+
+        String containerId = containerInfo.getContainerId();
+
+        containerInfo.getNetworkIds().forEach(networkId -> {
+            dockerClient.disconnectFromNetwork(containerId, networkId);
+            dockerClient.maybeRemoveNetwork(networkId);
+        });
+        dockerClient.stopAndRemoveContainer(containerId);
+    }
+
+    private static ExtensionContext.Store getStore(ExtensionContext context) {
+        return context.getStore(ExtensionContext.Namespace.GLOBAL);
```

Comment:
I didn't know about the store. Looks ok to me but I don't understand why you prefered this vs having a field in the class ?

Reply:
From what I understand its the recommended way to store state in extensions, and that extension classes should be stateless. It is essentially just a namespaced, hierarchical map, and having a map as a class field I imagine would perform in the exact same way. 
But apparently there are no guarantees about the exension lifecycle, so its better to use the store to store state. Plus it could be used by other extensions if they wanted to access the test state. 

https://blog.codefx.org/design/architecture/junit-5-extension-model/#Stateless

Reply:
Awesome ! Good catch !